File: backend/app/schemas/metrics.py
"""Schemas for training-history metrics (`GET /api/experiments/{id}/metrics`) and the leaderboard."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class MetricsResponse(BaseModel):
    """Per-episode training history for one experiment, chart-ready.

    `episodes` is row-oriented (one dict per episode, matching the source
    JSON almost exactly -- the shape most charting libraries, including
    Recharts, consume directly). `series` lists which metric keys are
    actually present for this agent type (DQN and PPO track different
    metrics), so a consumer doesn't have to guess or probe.
    """

    experiment_id: str
    agent: str
    total_episodes: int
    series: List[str] = Field(..., description="Metric keys present in every row of `episodes`.")
    episodes: List[Dict[str, Any]]


class LeaderboardEntry(BaseModel):
    """One row of `GET /api/leaderboard`."""

    agent: str
    win_rate: Optional[float] = None
    avg_episode_length: Optional[float] = None
    avg_reward: Optional[float] = None
    source: str = Field(
        ..., description="\"experiment_artifact\" (read live from rl/results/) or \"static_reference\" (README-recorded, no artifact exists)."
    )
    experiment_id: Optional[str] = Field(None, description="Which experiment this row's numbers came from, if artifact-sourced.")

File: frontend/src/types/metrics.ts
/** Mirrors `backend/app/schemas/metrics.py`. */

/**
 * One row of an experiment's training history. DQN and PPO log different
 * diagnostic fields (see `MetricsResponse.series` for which ones are
 * actually present on a given experiment) -- only the fields common to
 * both are typed explicitly; everything else is indexed loosely.
 */
export interface MetricPoint {
  episode: number;
  total_reward: number;
  steps: number;
  won: boolean;
  [metric: string]: number | boolean | string | null | undefined;
}

export interface MetricsResponse {
  experiment_id: string;
  agent: string;
  total_episodes: number;
  series: string[];
  episodes: MetricPoint[];
}

export interface LeaderboardEntry {
  agent: string;
  win_rate: number | null;
  avg_episode_length: number | null;
  avg_reward: number | null;
  /** At the default beginner/standard board: "experiment_artifact" (read
   * live from rl/results/) or "static_reference" (README-recorded, no
   * artifact exists). At any other level/density: "board_result" (read from
   * evaluate_board_config.py's output) or "not_trained" (no data yet). */
  source: "experiment_artifact" | "static_reference" | "board_result" | "not_trained";
  experiment_id: string | null;
}
